server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java

```diff
@@ -0,0 +1,237 @@
+            public void onNewClusterState(ClusterState state) {
+                logger.info("successfully removed all decommissioned nodes [{}] from the cluster", nodesToBeDecommissioned.toString());
+                nodesRemovedListener.onResponse(null);
+            }
+
+            @Override
+            public void onClusterServiceClose() {
+                logger.warn("cluster service closed while waiting for removal of decommissioned nodes.");
+            }
+
+            @Override
+            public void onTimeout(TimeValue timeout) {
+                logger.info("timed out while waiting for removal of decommissioned nodes");
+                nodesRemovedListener.onFailure(
+                    new OpenSearchTimeoutException(
+                        "timed out [{}] while waiting for removal of decommissioned nodes [{}] to take effect",
+                        timeout.toString(),
+                        nodesToBeDecommissioned.toString()
+                    )
+                );
+            }
+        }, allDecommissionedNodesRemovedPredicate);
+    }
+
+    public void updateMetadataWithDecommissionStatus(DecommissionStatus decommissionStatus, ActionListener<Void> listener) {
+        clusterService.submitStateUpdateTask(decommissionStatus.status(), new ClusterStateUpdateTask(Priority.URGENT) {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                Metadata metadata = currentState.metadata();
+                DecommissionAttributeMetadata decommissionAttributeMetadata = metadata.custom(DecommissionAttributeMetadata.TYPE);
+                assert decommissionAttributeMetadata != null && decommissionAttributeMetadata.decommissionAttribute() != null;
+                assert assertStatusTransitionOrFailed(decommissionAttributeMetadata.status(), decommissionStatus);
+                Metadata.Builder mdBuilder = Metadata.builder(metadata);
+                DecommissionAttributeMetadata newMetadata = decommissionAttributeMetadata.withUpdatedStatus(decommissionStatus);
+                mdBuilder.putCustom(DecommissionAttributeMetadata.TYPE, newMetadata);
+                return ClusterState.builder(currentState).metadata(mdBuilder).build();
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                DecommissionAttributeMetadata decommissionAttributeMetadata = newState.metadata()
+                    .custom(DecommissionAttributeMetadata.TYPE);
+                assert decommissionAttributeMetadata.status().equals(decommissionStatus);
+                listener.onResponse(null);
+            }
+        });
+    }
+
+    private static boolean assertStatusTransitionOrFailed(DecommissionStatus oldStatus, DecommissionStatus newStatus) {
```

Comment:
this method is not asserting anything, dont use "assert" in method name.

Reply:
changed name to validate and updated its called to - 
```
if(validateStatusTransitionOrFailed(decommissionAttributeMetadata.status(), decommissionStatus)) {
                    throw new DecommissioningFailedException(
                        decommissionAttributeMetadata.decommissionAttribute(),
                        "invalid previous decommission status found while updating status"
                    );
                }
```